lib/crash_reporter.js

```diff
@@ -0,0 +1,191 @@
+// SUBMIT_CRASH_URL = 'http://localhost:8080/crashes/submit';
+
+var preferenceFile = path.join(__dirname, '__preferences.json');
+
+var Preferences = function() {
+
+};
+
+Preferences.prototype.read = function(key, defaultValue) {
+  return this.readPreferences()
+    .then(function(contents) {
+      if (contents) {
+        return contents[key] || defaultValue;
+      } else {
+        return defaultValue;
+      }
+    })
+    .catch(function(error) {
+      logs.err('Error reading preference', key, error);
+      return defaultValue;
+    });
+};
+
+Preferences.prototype.write = function(key, value) {
+  // read existing preferences and write
+  var that = this;
+  var promise = new Promise(function(resolve, reject) {
+    that.readPreferences()
+      .then(function(contents) {
+        contents = contents || {};
+        contents[key] = value;
+        fs.writeFile(preferenceFile, JSON.stringify(contents), function(error) {
+          if (error) {
+            logs.err('Error writing preference', key, value);
+            reject(error);
+          } else {
+            resolve();
+          }
+        });
+      })
+      .catch(function(error) {
+        reject(error);
+      });
+  });
+  return promise;
+};
+
+Preferences.prototype.readPreferences = function() {
+  var promise = new Promise(function(resolve, reject) {
+    fs.exists(preferenceFile, function(exists) {
+      if (exists) {
+        fs.readFile(preferenceFile, function(error, data) {
+          if (error) {
+            reject(error);
+          } else {
+            resolve(JSON.parse(data));
+          }
+        });
+      } else {
+        // we don't have any local preferences
+        // return falsy value
+        resolve();
+      }
+    });
+  });
+  return promise;
+};
+
+// the only instance of the crash reporter
+var crashReporter = null;
+
+var CrashReporter = function() {
+  this.preferences = new Preferences();
+};
+
+CrashReporter.getInstance = function() {
+  if (crashReporter) {
+    return crashReporter;
+  } else {
+    crashReporter = new CrashReporter();
+    return crashReporter;
+  }
+};
+
+CrashReporter.prototype.turnOff = function() {
+  return this.preferences
+    .write(CRASH_REPORTER_PREFERENCE, 'off')
+    .catch(function(error) {
+      // do nothing
+      // do not crash the crash reporter :)
+      logs.err('Error turning off crash reporter preferences', error);
+    });
+};
+
+CrashReporter.prototype.turnOn = function() {
+  return this.preferences
+    .write(CRASH_REPORTER_PREFERENCE, 'on')
+    .catch(function(error) {
+      // do nothing
+      // do not crash the crash reporter :)
+      logs.err('Error turning on crash reporter preferences', error);
+    });
+};
+
+CrashReporter.prototype.submit = function(report) {
+  var that = this;
+  that.preferences.read(CRASH_REPORTER_PREFERENCE, 'on')
+    .then(function(value) {
+      if (value === 'on') {
+        var labels = [packageJson.name, packageJson.version].join(',');
```

Comment:
Other info that might be useful for a report:
* Operating System
* Node version
* npm version
* tessel version (although this might be harder to get from this point in the code)